### PR TITLE
Enabling pushing the log to KVP from the last pushed byte. 

### PR DESCRIPTION
**Pull request by Moustafa-Moustafa**
## Proposed Commit Message
Pushing the cloud-init.log file to the KVP store from the last pushed byte.

## Additional Context
Instead of pushing the logs only once, This change proposes enabling pushing the log multiple times without having to push any duplicate logs,

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

### CODE REVIEW

File: cloudinit/sources/helpers/azure.py [hunk 1]
```diff
@@ -214,26 +214,29 @@ def report_compressed_event(event_name, event_content):
...
+                                ": %s." % repr(ex), logger_func=LOG.debug)
 
     LOG.debug("Dumping cloud-init.log file to KVP")
     try:
         with open(file_name, "rb") as f:
             f.seek(0, os.SEEK_END)
-            seek_index = max(f.tell() - MAX_LOG_TO_KVP_LENGTH, 0)
+            seek_index = max(f.tell() - MAX_LOG_TO_KVP_LENGTH,
+                             last_log_byte_pushed_to_kvp_index)
             report_diagnostic_event(
```
Comment (johnsonshi) [inline cloudinit/sources/helpers/azure.py:233]: Report the value of `last_log_byte_pushed_to_kvp_index` so that we know the index being pushed.

**Comment by TheRealFalcon** [comment]
What is the goal of this change? Are you wanting a way to push additional log content if more gets added to the log after you've pushed it? Or if a log file is too large, do you want to be able to call it multiple times to push the whole thing? I think the fact that you're only taking the end of the log makes it a little confusing. Could you modify the docstring to clarify?

**Comment by Moustafa-Moustafa** [comment]
> What is the goal of this change? Are you wanting a way to push additional log content if more gets added to the log after you've pushed it? Or if a log file is too large, do you want to be able to call it multiple times to push the whole thing? I think the fact that you're only taking the end of the log makes it a little confusing. Could you modify the docstring to clarify?
This change has two main intentions:
1- Enabling pushing the log every time the machine is rebooted which will be helpful for different scenarios for Azure. This is achieved by moving the marker file to /run/cloud-init/. 
2- Continue pushing the log from the last pushed log byte, such that we push at most 500K bytes each time.

**Review by TheRealFalcon — APPROVED** [review]
Thanks for the clarification @Moustafa-Moustafa . I was expecting something that could push part of a file and the rest of the same file later...which is obviously different from what this is doing.

Code is +1 from me then, but can you add the clarification you added here on github into the docstring of `push_log_to_kvp`?

File: cloudinit/sources/helpers/azure.py [hunk 2]
```diff
@@ -214,32 +216,50 @@ def report_compressed_event(event_name, event_content):
...
     try:
         with open(file_name, "rb") as f:
             f.seek(0, os.SEEK_END)
-            seek_index = max(f.tell() - MAX_LOG_TO_KVP_LENGTH, 0)
+            seek_index = max(f.tell() - MAX_LOG_TO_KVP_LENGTH, start_index)
             report_diagnostic_event(
-                "Dumping last {} bytes of cloud-init.log file to KVP".format(
-                    f.tell() - seek_index),
+                "Dumping last {0} bytes of cloud-init.log file to KVP starting"
+                " from index: {1}".format(f.tell() - seek_index, seek_index),
```
Comment (OddBloke) [inline cloudinit/sources/helpers/azure.py:232]: A nit, but if you are substituting the variables in the order you specify them, then no need for the numbers:

```suggestion
                "Dumping last {} bytes of cloud-init.log file to KVP starting"
                " from index: {}".format(f.tell() - seek_index, seek_index),
```

**Review by OddBloke — APPROVED** [review]
Thanks Moustafa, this looks good to me! I have one comment inline, but it's a nit so there's no need to address it: just a note for next time.

**Pull request merged**